### PR TITLE
Patched mutagen not correctly handling corrupted ASF

### PR DESCRIPTION
**Pull request by LordSputnik**
Fixes the ASF/WMA issue in http://tickets.musicbrainz.org/browse/PICARD-770

### CODE REVIEW

File: picard/formats/asf.py [hunk 1]
```diff
@@ -23,9 +23,52 @@
...
+
+            self.size, self.num_objects = struct.unpack("<QL", header[16:28])
+            self.objects = []
+            for i in range(self.num_objects):
+                try:
+                    self.__read_object(fileobj)
+                except mutagen.asf.ASFError:
+                    break
+
+    ASF = DerivedASF
```
Comment (mineo) [inline picard/formats/asf.py]: DerivedASF != FixedASF.

**Comment by mineo** [comment]
Mutagen gets releases more often than Picard itself, could we just update the required version in `INSTALL.txt`?

I'm also a bit confused about [the commit in mutagen itself](https://bitbucket.org/lazka/mutagen/commits/2f2595228255d3d6038d7a2144189043) taking place in a completely different method.

**Comment by Sophist-UK** [comment]
I agree with @mineo - IMO this should NOT be a fix to Picard.

File: picard/formats/asf.py (hunk 1 above)
Comment (LordSputnik) [inline picard/formats/asf.py]: What do you mean by this?

**Comment by LordSputnik** [comment]
Picard uses 1.25 at the moment, which has a completely different asf module. So you would need to increase the minimum version requirement to 1.32. This is mainly a stopgap measure incase that's not possible for the next Picard release.

**Comment by Sophist-UK** [comment]
Actually Picard does not use a specific version of Mutagen. I suspect that:

a. If you are running from source code it uses whatever version you have installedI
b. If you are running from a Windows executable, it uses whatever version was installed on the environment which builds the executable.
c. If you are running on a Mac or Linux, I am not sure what happens.

So again I am with @mineo - we should move Picard executables to the latest version.

P.S. I am working on significant improvements to the Picard file format processing - and since I am testing this on an up to date mutagen, I will probably need the latest mutagen to make it work.

File: picard/formats/asf.py (hunk 1 above)
Comment (mineo) [inline picard/formats/asf.py]: The class you created is called `FixedASF`, but you're assigning `DerivedASF`.

Comment (LordSputnik) [inline picard/formats/asf.py]: Oh wow, I don't know how that ended up being different. Will fix.

**Comment by LordSputnik** [comment]
I've no objections to cancelling this and bumping the mutagen version :)

**Comment by phw** [comment]
> Picard uses 1.25 at the moment

As sophist explained above that's not a true statement per se. On Linux the system provided version is used. The OSX and Windows builds include a bundled version of Mutagen, and usually we use a recent version known to work. For Picard 1.3.2 that's indeed 1.25. 

So the right way to handle this would be to bump the version in the scripts [scripts/package-win.bat](https://github.com/musicbrainz/picard/blob/master/scripts/package-win.bat) and [scripts/package-osx.sh](https://github.com/musicbrainz/picard/blob/master/scripts/package-osx.sh) so we can test this in the daily builds. We should do this anyway to prepare for a new release. Not sure if we really need to document this as the minimum version as Picard will of course still work with earlier versions of mutagen.

Can the mutagen 1.32 version expected soon?

**Comment by zas** [comment]
Shouldn't we embed our own version of mutagen ? Since Picard and Mutagen are quite tied, it would help in a lot of cases imho. I'm not sure what it implies, but i think this is definitively something we could do, instead of "expecting" target system has a recent enough Mutagen version available.
Since Mutagen has no dependency but Python, having it packaged with Picard shouldn't much an issue.
What do you think about it ?

**Comment by phw** [comment]
> Shouldn't we embed our own version of mutagen ? Since Picard and Mutagen are quite tied, it would help in a lot of cases imho.

That would definitely be possible, and you are right it would solve some issues we had in the past. I usually have several mutagen versions in my Picard source folder for debugging mutagen related issues :) We already bundle mutagen for the OSX and Windows builds, but Linux distributions usually prefer using the system provided libraries. I wouldn't be surprised if some distributions would patch away the bundled mutagen version again.

But anyway, I am not opposed to the bundling as long as it gets automated in the build.

**Comment by Freso** [comment]
I wouldn't like Picard "forcibly" bundling a Mutagen version with it and would support Arch Linux unbundling it as part of its packaging. (Except I'm using picard-git, which I'm a co-maintainer of, so I'd vote in favour of us unbundling it there.)

**Comment by Sophist-UK** [comment]
I don't think we are talking about bundling a version of mutagen which would replace the system's default mutagen python package, but rather a private version that would only be used by Picard.

**Comment by Sophist-UK** [comment]
Another reason for Picard to include a local version of mutagen:

picard/formats/mutagenext/compatid3.py contains an override version of the mutagen.ID3 object, and changes in mutagen to the methods overwritten in compatid3.py may create incompatibilities.

If we are using a system-installed version of mutagen rather than our own local copy, then guaranteeing that compatid3 matches the mutagen version is impossible.

**Comment by phw** [comment]
That compatid3 code needs to be reviewed anyway because it is at least in parts obsolete.

**Comment by Sophist-UK** [comment]
I am already making changes to it. What needs reviewing?

**Comment by LordSputnik** [comment]
I believe one of the main reasons for having it at all was so that Picard could save to ID3v2.3. Since mutagen supports this now (since 1.26, I think?), it may be possible to get rid of it completely.

**Comment by Freso** [comment]
> I don't think we are talking about bundling a version of mutagen which would replace the system's default mutagen python package, but rather a private version that would only be used by Picard.

Which is exactly what I was arguing against.

**Comment by Sophist-UK** [comment]
@Freso - why are you against bundling it? It will ensure that Picard has a known version of mutagen to work with.

@LordSputnik - The changes I am making have a definite need for an alternative update_tov23 - I am not sure what the other parts of CompatID3 are for though. But it probably makes sense to use our update_tov23 to pre-process and then call the mutagen version. One other issue is a lack of ID3 override capability for AIFF files - no idea why it is not in mutagen when all other ID3 formats support it.

**Comment by Sophist-UK** [comment]
P.S. If anyone can look at the current version of CompatID3 and say which bits are still needed, I will make the changes as part of the work I am currently doing on format handling. (I am naturally very hesitant to make any changes to code the purpose of which I don't understand and which has not been explained in comments.)

**Comment by Sophist-UK** [comment]
P.P.S. The AIFF issues has been accepted by the Mutagen folks. Until then some tags will not be saved in AIFF ID3v23.

**Comment by Freso** [comment]
> @Freso - why are you against bundling it? It will ensure that Picard has a known version of mutagen to work with.

1. Picard releases less often than Mutagen so it would soon get out-of-date, lacking out on the benefits and bug fixes and feature additions of the more recent system-installed Mutagen.
2. It's never nice to keep duplicate code around on the system.

It would be much nicer to set a sensible minimum required version of Mutagen and update that, rather than forcing one specific instance at any time. The Mutagen folks are not introducing willy-nilly backwards incompatible changes.

**Comment by Sophist-UK** [comment]
@Freso - Agreed.

It seems to me that we might be able to get rid of CompatID3 altogether and just have a pre-processor for the standard update_tov23 which is in picard/formats/id3.py and which we call explicitly before calling update_tov23.

Please let me know if this approach is agreeable.

**Comment by lazka** [comment]
Some info on how to get rid of CompatID3: https://code.google.com/p/mutagen/issues/detail?id=85#c19

**Comment by zas** [comment]
> Picard releases less often than Mutagen so it would soon get out-of-date, lacking out on the benefits and bug fixes and feature additions of the more recent system-installed Mutagen.

You have a point here.
We could only use our bundled version of mutagen if the system's one is too old than the minimal version we require.
Also when a bug in Picard is related to Mutagen, we'd be able to do a bug fix release if it is something very annoying for our users.
Currently we can fix the bug in Picard, but we have no control if the issue comes (even partly) from Mutagen.

> It's never nice to keep duplicate code around on the system.

Right, but as a user i'd prefer to have duplicate code on my system and have a full featured app. Mutagen is quite light.

**Comment by lazka** [comment]
What about simply depending on the newest mutagen and thus forcing distros to update it?
Debian/Ubuntu does not allow bundling libraries, so imho this should be a last resort.

**Comment by mineo** [comment]
> We could only use our bundled version of mutagen if the system's one is older than the minimal version we require.

That only works if the new version is the one that fixes a bug, not the one that introduces it ;-)

If I understand the whole situation correctly, the version of mutagen used on Windows & Mac is completely under our control and doesn't change because we bundle it, which leaves only the Linux & BSD distributions where the mutagen version might not be what we expect. Those are usually run by nice people, I expect they'll make the necessary changes to their packages if we make it clear in our changelogs that we require a new version of mutagen.

**Comment by Sophist-UK** [comment]
Or for Linux/BSD we could check that the version of mutagen on the system is at least at our minimum version and warn users to upgrade at Picard startup if an upgrade is needed. This would be an easy fix.

**Comment by Sophist-UK** [comment]
@lazka - thanks for the pointer to the 2011-2013 issue. I will get my existing work on formats finished first, and submit it for review, and then see about removing compatID3 as a subsequent commit to the same PR #433 .

**Comment by Sophist-UK** [comment]
I have now reviewed the mutagen code for update_to23 and as far as I can see most of the code in CompatID3 is now superseded by the code in mutagen (which is based on this code anyway), and having it in Picard stops any subsequent fixes in mutagen being used. So IMO it should definitely go and I will include its removal in PR #433.

**Pull request closed without merging**